File: Labs/Lab4/Lab4.py
# Given the root of a binary tree, invert the tree, and return its root.


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def create_tree():
    root = TreeNode(1)

    # Left subtree
    root.left = TreeNode(2)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)

    # Right subtree
    root.right = TreeNode(8)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)

    return root


def invertTree(root):
    if root is None:
        return None

    if root:
        # Swap subtree roots
        root.left, root.right = root.right, root.left

        # Swap subtree leaves recursively
        invertTree(root.left)
        invertTree(root.right)

    return root


def print_tree(root):
    if root:
        # Print root
        print(root.val)

        # Print subtrees recursively
        print_tree(root.left)
        print_tree(root.right)


tree = create_tree()
print_tree(tree)
print("\n")

inverted_tree = invertTree(tree)
print_tree(inverted_tree)

File: Labs/Lab4/Lab4.java
public class Lab4 {
    public static void main (String[] args) {
        TreeNode tree = new TreeNode();
        tree = createTree();

        printTree(tree);
        System.out.println("\n");

        TreeNode invertedTree = new TreeNode();
        invertedTree = invertTree(tree);

        printTree(invertedTree);
    }

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public static TreeNode createTree() {
        TreeNode root = new TreeNode(1);

        // Left subtree
        root.left = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);

        // Right subtree
        root.right = new TreeNode(8);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(6);

        return root;
    }

    public static TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }
        else {
            // Swap subtree roots
            TreeNode temp = new TreeNode();
            temp = root.left;
            root.left = root.right;
            root.right = temp;

            // Swap subtree leaves recursively
            invertTree(root.left);
            invertTree(root.right);
        }

        return root;
    }

    public static void printTree(TreeNode root) {
        if (root != null) {
            // Print root
            System.out.println(root.val);

            // Print subtrees recursively
            printTree(root.left);
            printTree(root.right);
        }
    }
}
